src/algebra/periodic.lean

```diff
@@ -0,0 +1,264 @@
+/-
+Copyright (c) 2021 Benjamin Davidson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Benjamin Davidson
+-/
+import data.int.parity
+
+/-!
+# Periodicity
+
+In this file we define and then prove facts about periodic and antiperiodic functions.
+
+## Main definitions
+
+* `function.periodic`: A function `f` is *periodic* if `∀ x, f (x + c) = f x`.
+  `f` is referred to as periodic with period `c` or `c`-periodic.
+
+* `function.antiperiodic`: A function `f` is *antiperiodic* if `∀ x, f (x + c) = -f x`.
+  `f` is referred to as antiperiodic with antiperiod `c` or `c`-antiperiodic.
+
+Note that any `c`-antiperiodic function will necessarily also be `2*c`-periodic.
+
+## Tags
+
+period, periodic, periodicity, antiperiodic
+-/
+
+variables {α β γ : Type*} {f g : α → β} {c c₁ c₂ x : α}
+
+namespace function
+
+/-! ### Periodicity -/
+
+/-- A function `f` is said to be `periodic` with period `c` if for all `x`, `f (x + c) = f x`. -/
+@[simp] def periodic [has_add α] (f : α → β) (c : α) : Prop := ∀ x : α, f (x + c) = f x
+
+lemma periodic.comp [has_add α] (h : periodic f c) (g : β → γ) : periodic (g ∘ f) c :=
+by simp * at *
+
+@[to_additive]
+lemma periodic.mul [has_add α] [has_mul β] (hf : periodic f c) (hg : periodic g c) :
+  periodic (f * g) c :=
+by simp * at *
+
+@[to_additive]
+lemma periodic.div [has_add α] [has_div β] (hf : periodic f c) (hg : periodic g c) :
+  periodic (f / g) c :=
+by simp * at *
+
+lemma periodic.mul_const [division_ring α] (a : α) (h : periodic f c) :
+  periodic (λ x, f (x * a)) (c * a⁻¹) :=
```

Comment:
I think it might be worth also having
```lean
lemma periodic.const_smul {R} [group_with_zero R] [distrib_mul_action R α] (r : R) (h : periodic f c) :
  periodic (λ x, f (r • x)) (r⁻¹ • c) :=
```
from which you can get for free
```lemma
lemma periodic.const_mul [division_ring α] (a : α) (h : periodic f c) :
  periodic (λ x, f (a * x)) (a⁻¹ * c) := h.const_smul a
```

Reply:
Yes that's a good idea, thanks.

Reply:
In the end I added versions for `[module γ α]`, where `α` is an `add_comm_monoid` and `γ` is a `division_ring` (from which I also derive a version sing regular multiplication), and for `[distrib_mul_action γ α]`, where `α` is an `add_monoid` and `γ` is a `group`.